dll/win32/kernel32/client/sysinfo.c

```diff
@@ -423,21 +499,135 @@ SetFirmwareEnvironmentVariableA(IN LPCSTR lpName,
     return 0;
 }
 
-/*
- * @unimplemented
+/**
+ * @name EnumSystemFirmwareTables
+ * @implemented
+ *
+ * Obtains firmware table identifiers.
+ * https://msdn.microsoft.com/en-us/library/windows/desktop/ms724259(v=vs.85).aspx
+ *
+ * @param FirmwareTableProviderSignature
+ * Can be either ACPI, FIRM, or RSMB.
+ *
+ * @param pFirmwareTableBuffer
+ * Pointer to the output buffer, can be NULL.
+ *
+ * @param BufferSize
+ * Size of the output buffer.
+ *
+ * @return
+ * Actual size of the data in case of success, 0 otherwise.
+ *
+ * @remarks
+ * Data would be written to buffer only if the specified size is
+ * larger or equal to the actual size, in the other case Last Error
+ * value would be set to ERROR_INSUFFICIENT_BUFFER.
+ * In case of incorrect provider signature, Last Error value would be
+ * set to ERROR_INVALID_FUNCTION.
+ *
  */
 UINT
 WINAPI
 EnumSystemFirmwareTables(IN DWORD FirmwareTableProviderSignature,
                          OUT PVOID pFirmwareTableBuffer,
                          IN DWORD BufferSize)
 {
-    STUB;
-    return 0;
+    UINT uSize = 0;
+    UINT uCount = 0;
+    DWORD dwError = ERROR_SUCCESS;
+    PDWORD pBuffer = NULL;
+
+    switch (FirmwareTableProviderSignature)
+    {
+        case 'ACPI':
+        {
+            /* FIXME: Not implemented yet */
```

Comment:
Please return an appropriate error code here, not success.

Reply:
Done.